### DISCUSSION TITLE
Reading params deeply in Server Components (e.g. for i18n / multi-tenancy)

### DISCUSSION THREAD

**Discussion by amannn** [vercel/next.js#58862] [App Router]
Apps that use i18n or multi-tenancy are typically implemented with a top-level dynamic segment like `[locale]` or `[market]` (or a combination of both). Currently, to be able to use the value of these segments in Server Components, the only way is to [read them in a `page` or `layout` from props](https://nextjs.org/docs/app/building-your-application/routing/dynamic-routes#convention) and pass the values down to the components where they are needed (see e.g. [the Next.js i18n example](https://nextjs.org/docs/app/building-your-application/routing/internationalization)). 

While this can work for some apps, it can become quite cumbersome if many components are implemented as Server Components and require param values. My personal observation is that especially content-heavy apps can be implemented with a large number of components being Server Components. In the case of i18n, most components require the `locale` argument, therefore the developer would have to pass this value through almost all layers of components.

To mitigate this, i18n libraries like `next-intl` and `next-international` have introduced workarounds to pass a `locale` as a request header from the middleware to Server Components. While this improves the ergonomics, it opts all pages into dynamic rendering by default. Due to this, these libraries also provide `cache()`-based workarounds like [`setRequestLocale` in `next-intl`](https://next-intl-docs.vercel.app/docs/getting-started/app-router/with-i18n-routing#static-rendering) and [`setStaticParamsLocale` in `next-international`](https://next-international.vercel.app/docs/app-static-rendering#static-rendering) to be able to bring back static rendering capabilities. While this works, it's arguably a bit cumbersome too and requires special care from the developer.

I previously had some hopes that ServerContext could work for this use case, but [this API was removed from React](https://github.com/facebook/react/pull/27474). There was [a note](https://github.com/facebook/react/pull/27424#issuecomment-1739464985) from @sebmarkbage in the related PR that deprecated the feature though:

> There are perhaps other new features that frameworks can provide (like maybe Next.js can give access to part of the URL like i18n) that can be specific solutions to specific problems.

It would be very helpful if Next.js would provide a way to read the params that are passed as props to pages and layouts deeply from within the tree. The necessary constraint would be that only params from the current and parent segments can be read (same as the `params` prop).

**Example API:**

```tsx
// Same as `params` in the active segment
const {locale} = await params();
```

Related discussions:
- https://github.com/vercel/next.js/issues/43704
- https://github.com/vercel/next.js/discussions/51818
- https://github.com/vercel/next.js/issues/46618

Thank you for your consideration!

**Update:** Seems like this is coming in the form of [`next/root-params`](https://github.com/vercel/next.js/pull/80255/)!

**Comment by apavlinovic** [comment]
We have this exact use case, and it is cumbersome. It would be great if we could remove the `unstable_setRequestLocale` calls and let react-intl handle this. +1 on getting an official API to get the current route params server side.

**Comment by AhmedBaset** [comment]
Related: PR https://github.com/vercel/next.js/pull/59909 provides a new similar API `getPathname()`.

**Comment by stefanprobst** [comment]
@leerob can you clarify what the thoughts/plans of the team are in regards to this? thanks!

**Comment by AhmedBaset** [comment]
take a look at https://github.com/vordgi/next-impl-getters/blob/main/src/get-params.ts

I think it would be a better stopgap than our `setRequestLocale`

**Reply by amannn** [reply]
Interesting, thanks for sharing! It uses internals from Next.js, I'd be a bit worried about this breaking randomly between non-major Next.js versions (seems like [there were such changes already](https://github.com/vercel/next.js/issues/58244)), but it's interesting to see that this works.

**Reply by markomitranic** [reply]
@amannn interestingly they seem to be awaiting a PR merge into nextjs with that.

offtopic: More interestingly, their library seems to use `AsyncLocalStorage` for a server-context-like behaviour instead of `cache`. I wonder why and what is the difference. Do you happen to know?

**Reply by dferber90** [reply]
@markomitranic `cache` is not available in middleware, but AsyncLocalStorage is

**Reply by markomitranic** [reply]
Thanks for your answer @dferber90 !

I would be very careful with coupling the middleware/proxy execution to the layout or page render on a given machine or node process. There is a good reason why proxies only communicate through standardized systems - redirection, routing, url or header manipulation.

**Reply by thomas-j-wtfoxtrot** [reply]
Yes, for example, when deploying to Vercel, the middleware and the server rendering do not share memory. Here you need request headers or so.

**Comment by parsasabetz** [comment]


**Reply by markomitranic** [reply]
Hey @parsasabetz just for context, NextJS doesn't really have much to do with this, it is a React-level decision, modeled on how this same concept works in other server side languages. Tagging leerob isn't going to help you.

Additionally, stating that it needs to be "fixed" is a bit misleading, as nothing is broken, it was a concious decision made by the React team years ago.

What `next-intl` is doing, is perfectly in line with what any other server framework would do - You capture the locale from the pathname, then you store it in a shared global cache, and later you access it from the global scope.

The main difference between what `next-intl` does versus what Symfony or Rails would do is that they don't have 3 separate execution environments that don't share cache (middleware, layout, page) and which can be executed at completely separate times. Some server frameworks, like Elixir's Phoenix take it a step further and enforce prop drilling (or stateful process) instead of shared variables.

We are perfectly [capable to make server-side contexts](https://medium.com/homullus/cursed-server-context-patterns-in-next-js-14-64407c90fdd4) to prevent prop drilling and even share that data with the client components. We aren't, however, able to share logic, only data.

**Reply by eric-burel** [reply]
@markomitranic the article you linked is very interesting! Just a small nitpick
> We aren't, however, able to share logic, only data.

This is true for an isomorphic context that would be shared between client and server, which was the point of deprecated `createServerContext`. But for a pure server context based on the `cache` function, you can share "logic" in the form of non serializable objects. The first time I used this pattern was actually [to get proper GraphQL caching](https://prismic.io/blog/advanced-nextjs-server-context) using heavy clients like Apollo, with a request-scoped instance of the client being shared via cache.

**Comment by GabenGar** [comment]
My conclusion after dealing with internationalization and similar global-but-not-quite values is just accept prop drilling and pass them everywhere.
Any "convenient" solution would either chock in multi-environment/multi-threading/lambda setups or outright not be as convenient as it claims to be outside of hello world examples.
It's just an inherent problem of hybrid render since not only server and client have different globals, they also operate on different models. I.e. any client script at any given time is dealing with the context of one request, while the server constantly churns through them and renders context-dependant templates. Therefore any function accommodating both models will end up in a strict subset environment where it can't liberally use neither client nor server context. I also question the point of server components accessing route params deep in the render tree, since it creates an implicit dependency on the route which can be pretty annoying to debug and fix, especially if the component in question is overly generic (but not quite).

**Comment by amannn** [comment]
I had a quick look into what it'd take to implement this.

From my understanding, the RSC render starts here:

https://github.com/vercel/next.js/blob/0b261f0919c82c8b6f1ff08c10657ba9f57123ba/packages/next/src/server/app-render/app-render.tsx#L949-L959

As far as I understand, all layouts and the page that render as part of a request are all part of the `tree` that is passed to `ReactServerApp` here. This means that there is a single render call and the mechanism of rendering layouts and the page in parallel is handled by React internally.

Now the mechanism for "server context" in Next.js is based on `AsyncLocalStorage`. More specifically, the "context" is initialized here:

https://github.com/vercel/next.js/blob/0b261f0919c82c8b6f1ff08c10657ba9f57123ba/packages/next/src/server/app-render/app-render.tsx#L1504-L1510

`AsyncLocalStorage` works by providing a value deeply to a function closure. And that makes this a bit tricky to implement. 

Since there is only a single render call, I don't see an obvious place where `StaticGenerationAsyncStorageWrapper.wrap` could be called per rendering of layouts or pages. This is necessary though since layouts should only be able to access parent params, not potentially nested ones.

This poses the question: Should React provide a mechanism to distribute data via a context-like mechanism within Server Components since it "owns" the call to render components? The initial `createServerContext` proposal covered both the server as well as the client. From my experience, a server-only mechanism would be sufficient (I'd be happy with anything though).

I'm not really familiar with the Next.js codebase though, so if anyone has ideas or pointers, I'd be happy to discuss!

EDIT: Seems like `ReactFizzServer` has some traces of a request storage based on `AsyncLocalStorage` (see https://github.com/vercel/next.js/discussions/58862#discussioncomment-9506865)

**Reply by alexdln** [reply]
_Some additional context)_

As I already know from the support of createServerContext on AsyncLocalStorage (which I recently moved to a [separate package](https://github.com/vordgi/nimpl-context)) - it is impossible to make a stable server context on AsyncLocalStorage. It's strange to say such a thing to the author of the package, but yes.

Node.js has been holding this functionality unstable for 10 years already. And it is really unstable. When it is used in thousands of places, consistently 5 of them lose context. Often for the same pages, but not always, and sometimes they are all successfully assembled. In general, it is absolute randomness. And it is [described in the documentation [_Updated link_]](https://nodejs.org/dist/latest-v20.x/docs/api/async_context.html#troubleshooting-context-loss).

You can reduce the number of processes to one and then there will not be hundreds of instances living in parallel and the risks tend towards zero (_but it seems they are still there_). Next.js works over processes and thus controls their number. Although the likelihood of loss, as I understand, is still there (_but in such a situation next.js can simply restart the page build, although I did not delve into exactly how they circumvented this risk_).

**Reply by amannn** [reply]
Which part are you referring to in the docs where Node.js describes the feature as unstable? The docs on [Asynchronous context tracking in Node.js 20](https://nodejs.org/docs/latest-v20.x/api/async_context.html) declare the feature as stable:

<img width="478" alt="image" src="https://github.com/vercel/next.js/assets/4038316/fa3f69ac-7d72-409f-8ccc-134b01e4322c">

Individual features like `.bind` and `.snapshot` are declared as experimental, but e.g. `.run` and `.getStore` is stable.

**Reply by alexdln** [reply]
https://nodejs.org/dist/latest-v20.x/docs/api/async_context.html#troubleshooting-context-loss

**Reply by alexdln** [reply]
But yes, it is marked as stable now (_and has been stable for a long time_), I got confused somewhere. So it turns out it's stable, but with risks...

**Reply by eric-burel** [reply]
@vordgi since Next.js uses `AsyncLocalStorage` I think at worst, a solution based on it will be as robust as Next.js itself

@amannn digging the implementation of Next.js server context for `headers()/cookies()` is great but the issue is fundamentally a matter of architecture and framework design related to layouts, as the references you already quoted in your main post.

@delbaoliveira posted a response to the [Server Component + Layout.tsx - Can't access the URL / Pathname" dicussion](https://github.com/vercel/next.js/issues/43704#issuecomment-2090798307) giving more details on these choices. You can take a look at [my answer](https://github.com/vercel/next.js/issues/43704#issuecomment-2119143402) which raises a few more questions.

I've been working on i18n for the [Devographics (State of JS) surveys](https://github.com/Devographics/Monorepo/pull/367) lately, and ended up with an architecture similar to every other Next.js i18n frameworks for passing the locale param to RSC. It's convoluted, but the best we can do so far. 

As a bonus (unrelated to server context) we are introducing a token expression language that helps filtering the messages we send to the client, I am going to publish articles about that when it's a bit more polished.

**Reply by amannn** [reply]
Hey @eric-burel! Yep, I've seen that post by Delba, I absolutely understand the limitation in regard to the pathname since layouts might not re-render as part of RSC requests. For layout params I believe the situation is a bit different, where the limitation seems to not be conceptually, but on an implementation level.

I'm unsure what it would take from the technical perspective though to achieve this. Makes me wonder if `AsyncLocalStorage` should be renderable and consumable like `<SomeAsyncLocalStorage>{children}</SomeAsyncLocalStorage>` and `use(SomeAsyncLocalStorage)`? 🤔 

> As a bonus (unrelated to server context) we are introducing a token expression language that helps filtering the messages we send to the client, I am going to publish articles about that when it's a bit more polished.

Sounds interesting, would be curious to follow along! Is it compiler-based, similar to what is described in https://github.com/amannn/next-intl/issues/1#issuecomment-1497645889?

**Reply by alexdln** [reply]
> @vordgi since Next.js uses AsyncLocalStorage I think at worst, a solution based on it will be as robust as Next.js itself

Problems arise during assembly, when hundreds of contexts live in parallel in a short time. Next.js uses it over processes. This minimizes risks. But tried to find a reason why sometimes node.js can lose context, but I can't (_apparently the node.js developers can't find them too_).

That's why I implemented page-context logic ([PR](https://github.com/vercel/next.js/pull/65897)).

**Reply by alexdln** [reply]
> Makes me wonder if AsyncLocalStorage should be renderable and consumable like <SomeAsyncLocalStorage>{children}</SomeAsyncLocalStorage> and use(SomeAsyncLocalStorage)

That's what I did in [@nimpl/context](https://nimpl.tech/context/usage). You can try it, but in two projects where I tried to add this solution, it lost context with a probability of about 0.5%. As a result, I took it out into a separate package and removed it from my other packages.

**Reply by amannn** [reply]
Your implementation [with `storage.enterWith`](https://github.com/vordgi/nimpl-context/blob/292cefcb0e2be785f126ac874f5dd2a8e0333992/package/src/create-server-context.tsx#L8) is interesting … Makes me wonder if that could work, or if there's a risk that the context leaks out of a part of the tree, or as you've mentioned gets lost. If yes, this might require first-class support from React, but maybe worth digging a bit further into this. Another interesting case is when part of a tree suspends, if the suspended render calls still have access to the context.

There's also this note in [the docs](https://nodejs.org/api/async_context.html#asynclocalstorageenterwithstore):

> That is why run() should be preferred over enterWith() unless there are strong reasons to use the latter method.

If React would support this internally, maybe it could be implemented with `run()`. A userland solution is probably not able to use `run()`, since you only own the creation of React elements, not the call to actually render them.

**Reply by amannn** [reply]
As a side note, `ReactFizzServer` seems to already have [some usage of a request storage](https://github.com/facebook/react/blob/7621466b1b17965032aed24591e9b67fcd926a1b/packages/react-server/src/ReactFizzServer.js#L4313) (which is based on `AsyncLocalStorage`).

**Comment by alexdln** [comment]
Made a [PR](https://github.com/vercel/next.js/pull/65922) with the get-params implementation (_in the basic version - it's a very simple change_).

**Comment by eric-burel** [comment]
I am still digging this issue, and I feel like we would benefit from a kind of shared context, but that would allow different value client-side and server-side.
The shared part would specifically be the accessor function, but the context provider and the value stored in it could change depending on whether you are rendering server-side and client-side.

What led me to this idea:
- Server-side, we can carry the full dictionary around with barely any performance penalty. What we do with `cache`. But client-side we may want to filter only the tokens we use.
- We can have 2 different systems, fine, this works. Like a `ServerTrans` and a `ClientTrans` component to display messages.
- But then what about "shared" component, that are neither client components or server-components? You have no choice but to pick either `ServerTrans` or `ClientTrans`, usually you'd favour the latter since client components can render everywhere but server components can't. You can also pass translations via props from the parents but this defeats the purpose of having contexts.

API would be:
```jsx
import { getSharedIntlContext, SharedIntlContext } from "./Intl"

// A React Server Component layout
async function ServerLayout({children) {
const translations = await getTranslations()
const clientTranslations = filter(translations, clientTokens)
return (
  <div>
   {/* These translations are only available to server components */}
    <SharedIntlContext translations={translations}>
      <SomeServerComponent />
      <ClientLayout translations={clientTranslations} />
   </SharedIntlContext>
</div>
)
}

function ClientLayout({children, translations}) {
  // these translations are only available to client components
  return <SharedIntlContext translations={translations} children={children} />
}

// Display a translated message
// It will pick either the server context or the client context
// depending on where it's used
// There are 2 contexts, but only one shared accessor function
function SharedT({token}) {
  // sharedIntlContext() should work
  // client-side and server-side
  // but could return a different value
  // depending on the environment
  const { translate } = getSharedIntlContext()
  return <span>{translate(token)}</span>
}

```

The naming is not great but I hope you follow my point.

**Reply by amannn** [reply]
Note that you can use the `react-server` export condition to fork the implementation for Shared Components if you have a separate strategy for the server and client side (see [`useTranslations` in `next-intl`](https://next-intl-docs.vercel.app/docs/environments/server-client-components#rsc-background)).

**Reply by eric-burel** [reply]
That's a very smart trick, congrats on finding that. So I've tried to dig the code and check the documentation, it mentions that this shared approach works only for sync components. I guess you are expecting the messages to be available synchronously then? I see `.json` files in the docs so I guess you just import the messages at some point, using a synchronous `import`.

My problem is that the logic I've implemented that gets the messages in RSCs have to be async, because in the end it loads the messages using an API call. I could download the messages first hand and then build the app, Paraglide seems to do that too, but it's not great in a multi-tenant content where the available translation tokens may vary (for the State of JS 2023 and the State of CSS 2023, the tokens won't have the same keys). It's also non-idiomatic of frameworks that include data fetching logic exactly for this purpose, being able to render based on dynamic data.

Also you need a separate package to implement this strategy, not a blocker but having something built-in the framework would be better, even if the only use case ends up being implementation.

I think there is a progression margin in how React handle leaf components like translated text.

**Reply by markomitranic** [reply]
@eric-burel afaik the next-intl json import is an async action. I'm not an expert, but just looking at my own code:

```ts
export default getRequestConfig(async ({ locale }) => {
  if (!Object.keys(localeToLanguage).includes(locale)) {
    throw new Error(`Invalid locale ${locale} provided for translation file matcher.`);
  }

  return {
    messages: (
      await import(`../messages/${localeToLanguage[locale as keyof typeof localeToLanguage]}.json`)
    ).default,
  };
});
```

So i could easily fit a fetch call there if needed - but it only gets resolved once, during root layout generation. Sorry if this isn't what you were reffering to :D

**Reply by amannn** [reply]
@markomitranic Yep, that's correct!

Messages can be loaded asynchronously and from any data source. In case `useTranslations` is called in a Shared Component during the RSC render, the promise is resolved with the `use` hook and the component suspends until the messages are loaded. Note however that the promise result is cached, therefore this can happen at max. once per RSC render and can even have a slight perf benefit over async/await (see also [Should I use async or non-async functions for my components?](https://next-intl-docs.vercel.app/docs/environments/server-client-components#async-or-non-async)).

**Reply by amannn** [reply]
> but it's not great in a multi-tenant content where the available translation tokens may vary (for the State of JS 2023 and the State of CSS 2023, the tokens won't have the same keys)

Yep, good point! I agree there are valid cases for fetching translations dynamically!

> Also you need a separate package to implement this strategy, not a blocker but having something built-in the framework would be better, even if the only use case ends up being implementation.

Haven't tried it yet, but can you add a `package.json` within a regular folder in your project where you define the condition?

**Comment by sahandsn** [comment]
would love to see this resolved

**Comment by franky47** [comment]
There's a solution for this in [nuqs](https://nuqs.dev/docs/server-side) for search params, and it turns out it works just fine for params too. As a bonus point, you get type-safety for non-string params:

```tsx
// /blog/[date]/[slug]/page.tsx
import { 
  createSearchParamsCache as createParamsCache,
  parseAsIsoDateTime,
  parseAsString
} from 'nuqs/server'

type PageParams = {
  // Those are async in Next.js 15
  params: Promise<{ date: string, slug: string }>
}

const paramsCache = createParamsCache({
  date: parseAsIsoDateTime,
  slug: parseAsString
})

export default async function Page({ params }: PageProps) {
  paramsCache.parse(await params)
  return (
    <Deeply>
      <Nested>
        <ReactServerComponent />
      </Nested>
    </Deeply>
  )   
}

function ReactServerComponent() {
  // No prop drilling: reuse the same cache reference
  const slug = paramsCache.get('slug')
  const date = paramsCache.get('date')
  //    ?^ Date
}
```

**Reply by iDVB** [reply]
After reading this whole issue, this method makes complete sense to me. 
Seems like the most simple solution? Any issues limitations with it?
Should all pages use the same paramsCache instance? (Assuming you would create a single file that both your server components and page can reference vs in the same file as it is here?

**Reply by franky47** [reply]
The cache is only accessible on the server, for client components you would use the `useParams` hook (which isn't typed though).

**Comment by amannn** [comment]
As an update here, it seems like the Next.js team is working on a solution for this topic: [`rootParams`](https://github.com/vercel/next.js/pull/72837).

Sebastian Markbåge also shared on Twitter that cache-based patterns that need to be initialized via a setter will likely break with upcoming features in Next.js (the post was later deleted, as he closed his account). Therefore once `rootParams` is available, I think it'll be a good idea to transition to it.

**Reply by franky47** [reply]
Thanks for the pointer, rootParams looks like a better alternative to nuqs' cache hack (which would be impacted by Sebastian's announced behaviour change).

cc @iDVB

**Reply by daveycodez** [reply]
Looks like this is the blocker for localized static applications using the App Router. Will stick to Pages Router until this is released and properly documented in an updated next-intl

**Reply by yafkari** [reply]
In case anyone would like to be notified, the PR mentionned has been merged

**Reply by omarkhatibgg** [reply]
@yafkari I tried it on canary, it did not work with me 🥲
